CORE-18534 External Processors to return transient retry codes

conal-smith opened the pull request:
Note - This is the first PR in a series of PRs to update sync-RPC processors on the server side. It includes:
- messaging layer changes, `CordaHTTPTransientServerException` and handling to return HTTP response with 503 status code 
- `EntityRequestProcessor` changes

**Problem description:**
All flow engine external events go through the RPCClient in the messaging library. This has a built-in retry mechanism to retry events for some period of time. If an external event hits a transient error in the external event handler, the flow engine should rely on this retry handling rather than implementing its own.

To do this, the external event handlers must always return a retry code if a transient error is encountered. It may be possible to do this easily using the flow external event library that these handlers are currently using, or it may require some adjustments to the messaging library code to allow for this case.

Completed:
- [x] `EntityRequestProcessor` 

Remaining (separate PRs):
- [ ]  `CryptoFlowOpsProcessor` 
- [ ] `LedgerPersistenceRequestProcessor` 
- [ ] `TokenSelectionSyncRPCProcessor` 
- [ ] `UniquenessCheckMessageProcessor` 
- [ ] `VerificationRequestProcessor`

conal-smith committed:
CORE-18534 - rename `times` to `maxRetries`

conal-smith committed:
CORE-18534 - add new transient exception type and handling to messaging library and impl.
- utilize the new transient exception handling in the persistence service

conal-smith committed:
CORE-18534 - remove irrelevant refactoring

conal-smith committed:
CORE-18534 - extract exception handling to a util and test it

conal-smith committed:
CORE-18534 - rename the RPC API exception file, rename the transient HTTP server exception

sprintengine commented on libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/exception/CordaRPCAPIExceptions.kt:
```diff
@@ -19,15 +19,3 @@ class CordaRPCAPIPartitionException(message: String?, exception: Throwable? = nu
  */
 class CordaRPCAPIResponderException(val errorType: String, message: String?, exception: Throwable? = null) :
     CordaRuntimeException(message, exception)
-
```
This file has exceptions used by the RPC send/receive message pattern. I did not want to mix the new exception types used in sync rpc/http message pattern so I renamed this back to "CordaRPCAPIExceptions" and created a new CordaHTTPExceptions containing client- and server-side exceptions for the sync rpc/http pattern.

conal-smith committed:
CORE-18534 - update entity persistence exceptions for transient exception being thrown

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/SyncRPCSubscriptionImpl.kt:
```diff
@@ -90,11 +91,7 @@ internal class SyncRPCSubscriptionImpl<REQUEST : Any, RESPONSE : Any>(
-                    context.result(errorMsg)
-                    context.status(ResponseCode.INTERNAL_SERVER_ERROR)
-                    return@trace context
+                    return@trace handleProcessorException(log, endpoint, ex, context)
```
messaging code - handling transient exceptions here. I think this is more suitable than an exception handler registered into the Javalin webserver.

sprintengine commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/SyncRPCUtils.kt:
```diff
@@ -0,0 +1,33 @@
+                log.warn(msg, ex)
+                context.result(msg)
+            }
+            context.status(ResponseCode.SERVICE_UNAVAILABLE)
```
How do you feel about a 503 for temporary transient errors on external processors?

JamesHR3 replied:
I think that's probably the right option here.

jujoramos commented on components/persistence/persistence-service-common/src/main/kotlin/net/corda/persistence/common/ResponseFactoryImpl.kt:
```diff
@@ -44,7 +45,7 @@ class ResponseFactoryImpl @VisibleForTesting internal constructor(
     override fun errorResponse(externalEventContext : ExternalEventContext, exception: Exception) = when (exception) {
         is CpkNotAvailableException, is VirtualNodeException -> {
-            transientErrorResponse(externalEventContext, exception)
+            throw CordaHTTPServerTransientException(externalEventContext.requestId, exception)
```
Should we keep the warning message (previously logged within `transientErrorResponse() -> errorLogMessage()`)?.

jujoramos commented on components/persistence/persistence-service-common/src/main/kotlin/net/corda/persistence/common/ResponseFactoryImpl.kt:
```diff
@@ -56,22 +57,14 @@ class ResponseFactoryImpl @VisibleForTesting internal constructor(
                 PersistenceExceptionType.FATAL -> fatalErrorResponse(externalEventContext, exception)
                 PersistenceExceptionType.PLATFORM -> platformErrorResponse(externalEventContext, exception)
-                PersistenceExceptionType.TRANSIENT -> transientErrorResponse(externalEventContext, exception)
+                PersistenceExceptionType.TRANSIENT -> throw CordaHTTPServerTransientException(externalEventContext.requestId, exception)
```
Should we keep the warning message (previously logged within `transientErrorResponse() -> errorLogMessage()`)?.

sprintengine replied:
The original message read:
```
Exception occurred (type=$errorType) for flow-worker request ${flowExternalEventContext.requestId}
```

The messaging code in SyncRPCSubscriptionImpl logs a warning that reads:
```
Transient error processing RPC request for $endpoint: Transient server exception while processing request '$requestId'. Cause: ${cause?.message}
```

Not sure if we need both?

jujoramos replied:
I didn't notice we were logging the warning in other place 👍

jujoramos commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/SyncRPCUtils.kt:
```diff
@@ -0,0 +1,33 @@
+package net.corda.messaging.utils
```
Can we use `HTTP` instead of `RPC` for names, exception messages, logging, etc. (across the whole `PR`, not just this file)?.

sprintengine replied:
Are you suggesting rename `SyncRPCProcessor` to `SyncHTTPProcessor`? I wouldn't be against that, maybe in a follow up PR however. Following James' comment below I removed this utils file anyway.

jujoramos replied:
👍

JamesHR3 commented on libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/processor/SyncRPCProcessor.kt:
```diff
@@ -13,6 +13,7 @@ interface SyncRPCProcessor<REQUEST, RESPONSE> {
      * @param request
      *
      * @return the result of the processing of the type RESPONSE
+     * @throws [CordaTransientServerException] if a transient exception occurred that can be retried by the client.
```
I don't know that the kdocs in this interface do a great job of describing what a user of the messaging library must do to handle particular cases. I think here for example we should state clearly that if the processing throws a `CordaTransientServerException`, then the messaging library will go into retry loop in which the `process` method is retried with backoff. Clients therefore need to make sure that exception is thrown if an encountered error is transient.

As an aside, we should also state that implementors need to be idempotent.

sprintengine replied:
> if the processing throws a CordaTransientServerException, then the messaging library will go into retry loop in which the process method is retried with backoff

The messaging library won't perform the retry, it translates the transient exception to a 503 status and lets the HTTP client perform retry with backoff. 

I've made a few updates to the kdocs, let me know what you think?

JamesHR3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/utils/SyncRPCUtils.kt:
```diff
@@ -0,0 +1,33 @@
+import net.corda.web.api.WebContext
+import org.slf4j.Logger
+
+fun handleProcessorException(
```
Shouldn't this be part of the sync RPC subscription class? I don't think we reuse it anywhere?

sprintengine replied:
My main concern was really just splitting the exception handling so I could test it in isolation but it's probably not necessary, I'll chuck it back into the messaging class

conal-smith committed:
CORE-18534 - remove utils class and test, add tests to subscription

conal-smith committed:
CORE-18534 - align kdocs with updates
- mention idempotency in SyncRPCProcessor
- move some RPCSubscriptionImpl specific kdocs from the RPCSubscription interface, which now has two implementations
- add more explicit documentation about throwing the transient exception

jujoramos reviewed (approved):
LGTM

kyriathar reviewed (approved):
OK (db changes)

LWogan reviewed (approved):
LGTM - flow
